README
Sys/MemInfo version 0.98
========================

This module return the total amount of free and used physical memory
in bytes in totalmem and freemem variables.

This module has been tested on Linux 3.13.0, UnixWare 7.1.2, AIX5, OpenBSD 3.8, 
NetBSD 2.0.2, FreBSD 5.4, HPUX11, Solaris 9, Tru64 5.1, Irix 6.5, MacOS X 10.2
and Windows XP.

It should work on FreeBSD 4 and Windows 9X/ME/NT/200X/Vista.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  File::Copy

COPYRIGHT AND LICENCE

Copyright (C) 2005, 2006, 2014 Sylvain Cresto <scresto _at_ gmail.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 


Changes
Revision history for Perl extension Sys::MemInfo.

0.01  Thu Oct 24 01:10:47 2005
    - original version

0.02  Fri Dec 02 03:08:00 2005
    - added HPUX11 support

0.03  Tue Dec 06 02:36:00 2005
    - added WinXP and Windows Server 2003 support
    - added real test, thanks H. Merijn Brand

0.04  Wed Dec 21 02:08:00 2005
    - added Tru64 support

0.05  Wed Aug 16 23:14:32 2006
    - fixed wrong memory report (long overflow on linux)
      thanks Laurent Dufour <dufour_l _at_ hotmail.com>

0.90  Sun Sep 17 10:24:59 2006
    - added win 9x/ME/NT support
    - use GlobalMemoryStatusEx with WinXP, no longuer
      depend on psapi.dll
    - added total and free swap informations (totalswap
      and freeswap directives)
    - added availkeys and get functions,
      thanks Wilson Snyder <wsnyder _at_ wsynder.org>
    - debug *bsd support
    - added Irix support

0.91  Thu Nov 09 23:49:00 2006
    - added Mac OS X Darwin support
    - debug freebsd support
    - patch test script, thanks Wilson Snyder

0.98  Wed Sep 14 22:55:00 2014
    - added cygwin support (RT #85152), thanks <kqfhjjgrn _at_ yahoo.es>
    - RT #80636
    - fixed pod abstract (RT #94160), thanks <NEILB _at_ cpan.org>


Makefile.PL
use 5.006;
use ExtUtils::MakeMaker qw(prompt WriteMakefile);
use File::Copy 'copy';

my $libs = '';
my $define = '';

for ($^O) {
  if (/svr/) {
    copy ('arch/sco.xs', 'MemInfo.xs');
    print "Sys::MemInfo for UnixWare\n";
  } elsif (/aix/) {
    copy ('arch/aix.xs', 'MemInfo.xs');
    $libs = '-lperfstat';
    print "Sys::MemInfo for AIX\n";
  } elsif (/linux/ or /cygwin/) {
    copy ('arch/linux.xs', 'MemInfo.xs');
    my $kver = `uname -r`;
    die "Error extracting kernel version.\n" unless ($kver =~ /^(\d+)\.(\d+)\.(\d+)/);
    if ($1<=2 && ($2<3 || ($2==3 && $3<17))) {
      $define = '-DOLDKERNEL';
      print "Sys::MemInfo for Linux 2.3.16 and lower\n";
    } else {
      print "Sys::MemInfo for Linux 2.3.17 and higher\n";
    }
  } elsif (/solaris/) {
    copy ('arch/solaris.xs', 'MemInfo.xs');
    print "Sys::MemInfo for Solaris\n";
  } elsif (/freebsd/) {
    copy ('arch/freebsd.xs', 'MemInfo.xs');
    $kver = `uname -r`;
    die "Error extracting kernel version.\n" unless ($kver =~ /^(\d+)\./);
    if ($1>=5) {
      $define = '-DFREEBSD5';
      print "Sys::MemInfo for FreeBSD 5 and higher\n";
    } else {
      print "Sys::MemInfo for FreeBSD 4 and lower\n";
    }
  } elsif (/bsd/) {
    copy ('arch/bsd.xs', 'MemInfo.xs');
    if (/netbsd/) {
      $define = '-DNETBSD';
      print "Sys::MemInfo for NetBSD\n";
    } else {
      print "Sys::MemInfo for *BSD\n";
    }
  } elsif (/hpux/) {
    copy ('arch/hpux.xs', 'MemInfo.xs');
    $define ='-D_XOPEN_SOURCE_EXTENDED';
    print "Sys::MemInfo for HPUX\n";
  } elsif (/dec_osf/) {
    copy ('arch/tru64.xs', 'MemInfo.xs');
    $libs='-lmach';
    print "Sys::MemInfo for Tru64\n";
  } elsif (/MSWin32/) {
    copy('arch/win32.xs', 'MemInfo.xs');
    require Win32;	
    my (undef, $version)= Win32::GetOSVersion() ;
    if ($version && 5 <= $version) {
      print "Sys::MemInfo for Windows 200X/XP/Vista and higher\n";
    } else {
      print "Sys::MemInfo for Windows 9X/ME/NT\n";
      $define='-DOLDWIN';
    }  
  } elsif (/irix/) {
    copy('arch/irix.xs', 'MemInfo.xs');
    print "Sys::MemInfo for Irix\n";
  } elsif (/darwin/) {
    copy('arch/darwin.xs', 'MemInfo.xs');
    print "Sys::MemInfo for MacOS X Darwin\n";
  } else {
    die "unsupported OS: $_\n";
  }
}


WriteMakefile(
  'NAME'         => 'Sys::MemInfo',
  'VERSION_FROM' => 'MemInfo.pm', # finds $VERSION
  'DEFINE'       => $define,
  'NORECURS'     => 1,
  'XSOPT'        => '-noprototypes',
  'XS'           => { 'MemInfo.xs'  => 'MemInfo.c', },
  clean => { 'FILES' => 'MemInfo.xs', },
  'LIBS'         => $libs,

  ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
    (ABSTRACT_FROM => 'MemInfo.pm', # retrieve abstract from module
    'AUTHOR'       => 'Sylvain CRESTO <scresto@gmail.com>') : ()),
);

# vim:et:ts=2:sts=2:sw=2


MemInfo.pm
package Sys::MemInfo;

use strict;
use warnings;

require Exporter;
require DynaLoader;

our @ISA = qw(Exporter DynaLoader);
# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.
our @EXPORT = qw(
);
our @EXPORT_OK = qw(
totalmem
freemem
totalswap
freeswap
availkeys
get
);
our $VERSION = 0.98;
bootstrap Sys::MemInfo $VERSION;

sub get {
  my $field = shift;
  no strict;
  return &{$field};
}

=head1 NAME

Sys::MemInfo - query the total free and used physical memory

=head1 SYNOPSIS

  use Sys::MemInfo qw(totalmem freemem totalswap);

  print "total memory: ".(&totalmem / 1024)."\n";
  print "free memory:  ".(&freemem / 1024)."\n";

  print "total swap: ".(&totalswap / 1024)."\n";
  print "free swap:  ".(Sys::MemInfo::get("freeswap") / 1024)."\n";

=head1 DESCRIPTION

Sys::MemInfo return the total amount of free and used physical memory in bytes in totalmem and freemem variables.

Total amount of free and user swap memory are alse returned in totalswap and freeswap variables.

This module has been tested on Linux 3.13.0, UnixWare 7.1.2, AIX5, OpenBSD 3.8, 
NetBSD 2.0.2, FreBSD 5.4, HPUX11, Solaris 9, Tru64 5.1, Irix 6.5, Mac OS X 10.2 darwin and Windows XP.

It should work on FreeBSD 4 and Windows 9X/ME/NT/200X/Vista.

=head1 METHODS

=over 4

=item availkeys

Return list of all accessor keys (freemem, totalmem, etc.)  This is useful
for dumping out all known information from the object by calling get() on
all of the returned keys.

=item freemem

Returns free physical memory in bytes.

=item freeswap

Returns free swap space in bytes.

=item get

Returns the value of the passed key.

=item totalmem

Returns total physical memory size in bytes.

=item totalswap

Returns total swap space in bytes.

=back

=head1 AUTHOR

Sylvain Cresto E<lt>scresto@gmail.comE<gt>

Thanks to Laurent Dufour and Wilson Snyder.

=head1 BUGS

Please send bug-reports to scresto@gmail.com

=head1 LICENCE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=head1 COPYRIGHT

Copyright (C) 2005, 2006, 2014 - Sylvain Cresto

=cut

1;
